### PR TITLE
[WIP] cache for dataloader

### PR DESCRIPTION
**Pull request by k2kobayashi**
- Cache sample for next epoch

### CODE REVIEW

**Comment by k2kobayashi** [comment]
It may affect conversion quality because this implementation fixes ```cv_spkr_name``` in each sample while training.

**Comment by unilight** [comment]
> It may affect conversion quality because this implementation fixes `cv_spkr_name` in each sample while training.

Could you explain a little bit more about this part? What was the situation in the original implementation?

**Comment by k2kobayashi** [comment]
Sorry, I just memorized. In this function, ```cv_spkr_name``` is sampled when a single-batch is loaded. However, when it uses cache, the sampling in each batch is performed only once over whole training. https://github.com/k2kobayashi/crank/blob/bba07f909257f195313ebf1fecf319bc2e52aaeb/crank/net/trainer/dataset.py#L70

**Comment by unilight** [comment]
> Sorry, I just memorized. In this function, `cv_spkr_name` is sampled when a single-batch is loaded. However, when it uses cache, the sampling in each batch is performed only once over whole training.

I think I understand, but since I am not really familiar with Pytorch's DataLoader, let me confirm. Using this new function, all the processes (loading, pre-/mid-/post-processing, generating random index) will be done for the first epoch only. Is that right? I think this is no problem as long as `shuffle=True`, as in
https://github.com/k2kobayashi/crank/blob/bba07f909257f195313ebf1fecf319bc2e52aaeb/crank/net/trainer/utils.py#L118

**Review by unilight — APPROVED** [review]

**Comment by k2kobayashi** [comment]
Yes, it's by this implementation. In typical Pytorch dataloader, the all processes is done in each epoch. The argument ```shuffle``` means the ```idx``` is selected in random order. https://github.com/k2kobayashi/crank/blob/08998d6894572302ac0935af2455f298cfc770b2/crank/net/trainer/dataset.py#L47

**Comment by unilight** [comment]
I see. LGTM.
